components/veml7700/veml7700.c

```diff
@@ -0,0 +1,237 @@
+    do                                                                                                                 \
+    {                                                                                                                  \
+        if (!(VAL))                                                                                                    \
+            return ESP_ERR_INVALID_ARG;                                                                                \
+    }                                                                                                                  \
+    while (0)
+
+static esp_err_t read_port(i2c_dev_t *dev, uint8_t command_code, uint16_t *data)
+{
+    CHECK_ARG(dev);
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_read(dev, &command_code, 1, data, 2));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+static esp_err_t write_port(i2c_dev_t *dev, uint8_t command_code, uint16_t data)
+{
+    CHECK_ARG(dev);
+
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_write(dev, &command_code, 1, &data, 2));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+static uint32_t resolution(veml7700_config_t *config)
+{
+    CHECK_ARG(config);
+
+    uint32_t resolution = VEML7700_RESOLUTION_800MS_IT_GAIN_2;
+    switch (config->gain)
+    {
+        case VEML7700_GAIN_1:
+            resolution = resolution * 2;
+            break;
+        case VEML7700_GAIN_DIV_4:
+            resolution = resolution * 8;
+            break;
+        case VEML7700_GAIN_DIV_8:
+            resolution = resolution * 16;
+            break;
+    }
+    switch (config->integration_time)
+    {
+        case VEML7700_INTEGRATION_TIME_400MS:
+            resolution = resolution * 2;
+            break;
+        case VEML7700_INTEGRATION_TIME_200MS:
+            resolution = resolution * 4;
+            break;
+        case VEML7700_INTEGRATION_TIME_100MS:
+            resolution = resolution * 8;
+            break;
+        case VEML7700_INTEGRATION_TIME_50MS:
+            resolution = resolution * 16;
+            break;
+        case VEML7700_INTEGRATION_TIME_25MS:
+            resolution = resolution * 32;
+            break;
+    }
+    return resolution;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+esp_err_t veml7700_init_desc(i2c_dev_t *dev, i2c_port_t port, gpio_num_t sda_gpio, gpio_num_t scl_gpio)
+{
+    CHECK_ARG(dev);
+
+    dev->port = port;
+    dev->addr = VEML7700_I2C_ADDR;
+    dev->cfg.sda_io_num = sda_gpio;
+    dev->cfg.scl_io_num = scl_gpio;
+#if HELPER_TARGET_IS_ESP32
+    dev->cfg.master.clk_speed = I2C_FREQ_HZ;
+#endif
+
+    return i2c_dev_create_mutex(dev);
+}
+
+esp_err_t veml7700_free_desc(i2c_dev_t *dev)
+{
+    CHECK_ARG(dev);
+
+    return i2c_dev_delete_mutex(dev);
+}
+
+esp_err_t veml7700_probe(i2c_dev_t *dev)
```

Comment:
Why is this function needed at all? Isn't it enough for the user to call `i2c_dev_probe()`?

Reply:
I'd like to keep the probe function, it might me enough to use the i2c_dev_probe if the write operation is used. The probe does not work with a read operation, the chip doen't ack when the command byte write is missing before a read operation. Having the probe function in the driver prevents making this mistake (again).